### PR TITLE
OAK-12406 IF and EXISTS functions for sparse function-based indexing

### PR DESCRIPTION
**Pull request by thomasmueller**

### CODE REVIEW

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/FunctionOperatorUtils.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            return null;
+        }
+    }
+
+    /**
+     * Compare two (non-null) operands: numerically if both can be parsed as
+     * a number, otherwise as strings.
+     */
+    static int compareOperands(PropertyValue a, PropertyValue b) {
+        Double da = toDouble(a);
```
Comment (bhabegger) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/FunctionOperatorUtils.java:81]: Systematic attempt to convert to double ? Is there not a way to know the type of a PropertyValue ?

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/KeysetPaginationTest.java [hunk 2]
```diff
@@ -0,0 +1,301 @@
+/*
```
Comment (bhabegger) [inline oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/KeysetPaginationTest.java:1]: Is this within the scope of this PR ? What does keyset pagination have to do with the introduction of if an exists ?

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/ExistsImpl.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+import org.apache.jackrabbit.oak.plugins.memory.PropertyValues;
+import org.apache.jackrabbit.oak.query.index.FilterImpl;
+import org.apache.jackrabbit.oak.spi.query.QueryConstants;
+import org.apache.jackrabbit.oak.spi.query.QueryIndex.OrderEntry;
+
+/**
+ * The function "exists(..)". Returns true if the operand has a value, and
+ * false otherwise. Always determinate (never returns null).
+ */
+public class ExistsImpl extends DynamicOperandImpl {
```
Comment (bhabegger) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/ExistsImpl.java:37]: Where is this logic tested ?

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/IfImpl.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+import org.apache.jackrabbit.oak.api.Type;
+import org.apache.jackrabbit.oak.commons.collections.SetUtils;
+import org.apache.jackrabbit.oak.query.index.FilterImpl;
+import org.apache.jackrabbit.oak.spi.query.QueryConstants;
+import org.apache.jackrabbit.oak.spi.query.QueryIndex.OrderEntry;
+
+/**
+ * The function "if(condition, trueValue, falseValue)".
+ */
+public class IfImpl extends DynamicOperandImpl {
```
Comment (bhabegger) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/IfImpl.java:36]: Where is this logic tested ?

File: oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/FunctionIndexCommonTest.java [hunk 5]
```diff
@@ -969,6 +969,78 @@ public void coalesce() throws Exception {
...
+        a.setProperty("f", "no");
+        Tree b = rootTree.addChild("b");
+        b.setProperty("cond", false);
+        b.setProperty("t", "yes2");
+        b.setProperty("f", "no2");
+
+        root.commit();
+
+        assertEventually(() -> assertPlanAndQuery(
+                "select * from [nt:base] where if([cond], [t], [f]) = 'yes'",
```
Comment (bhabegger) [inline oak-search/src/test/java/org/apache/jackrabbit/oak/plugins/index/FunctionIndexCommonTest.java:1020]: This might be a bit light if it's the only testing we have for the if construct (happy path).

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/FulltextRegexQueryTest.java [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+ * {@link #fuzzForEncodingThatReachesGenuineRegexParsing}). A {@code native('lucene', ...)}
+ * clause, however, bypasses that escaping and does genuinely reach
+ * {@code RegExp.toAutomaton()} (see {@link #nativeLuceneQueryReachesGenuineRegexParsing}
+ * and {@link #pathologicalRegexCanCauseCatastrophicSlowdown}).
+ * <p>
+ * Note oak-lucene bundles its own (very old, Lucene 4.7.2 based, see OAK-10786) copy of
+ * the automaton classes, so this needs to be verified against oak-lucene's actual runtime
+ * behaviour rather than assumed from the upstream Lucene ticket.
+ */
+public class FulltextRegexQueryTest {
```
Comment (bhabegger) [inline oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/FulltextRegexQueryTest.java]: Is this related to the introduction of IF and EXISTS ?

File: oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/util/FunctionIndexProcessor.java
Comment (bhabegger) [inline oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/util/FunctionIndexProcessor.java:121]: I think having some tests for this function would help understand the intent and validate correctness. 
All these ifs in a for loop is an open door for edge case errors.

File: oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/util/FunctionIndexProcessor.java [hunk 7]
```diff
@@ -105,11 +122,59 @@ public static String[] getFunctionCode(String functionDescription) {
...
+        if ("op".equals(functionName)) {
+            // op(a, operator, b); operator is typically a quoted string
+            // literal, but can be any operand that evaluates to one of the
+            // known operator strings (e.g. a property reference)
+            PropertyState a = stack.pop();
+            PropertyState operator = stack.pop();
+            PropertyState b = stack.pop();
+            return calculateOp(a, operator, b);
+        }
         PropertyState ps = stack.pop();
```
Comment (bhabegger) [inline oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/util/FunctionIndexProcessor.java:178]: pop here before "coalesce" and not before "if" seems surprising...

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneFunctionIndexCommonTest.java [hunk 8]
```diff
@@ -86,4 +91,44 @@ public void shutdownExecutor() {
...
     }
 
+    /**
+     * An index whose only indexed property is a function that evaluates to
+     * null for most nodes must stay "sparse": it should contain a Lucene
+     * document only for the nodes where the function is not null, not one
+     * document per node.
+     */
+    @Test
+    public void sparseIndexForIfExistsFunction() throws CommitFailedException {
```
Comment (bhabegger) [inline oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneFunctionIndexCommonTest.java:101]: Did we introduce index time if also (as a side effect ?) Or only query time ?

Comment (thomasmueller) [reply oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/LuceneFunctionIndexCommonTest.java:101]: For all functions, we need to implement them both at indexing time (even the parsing of functions, because they need to be in the index definition) as well as query time (parsing the XPath + SQL-2, and to send the function to the indexes).

File: oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/util/FunctionIndexProcessor.java (hunk 7 above)
Comment (thomasmueller) [reply oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/util/FunctionIndexProcessor.java:178]: All functions pop at least one operand. Maybe it's easier to understand if there is one switch depending on the function name, and then each function is implemented separately (it would be more code thought, because we would have a lot of repetition).

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/ExistsImpl.java (hunk 3 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/ExistsImpl.java:37]: In SQL2ParserTest, for example... I'll look at the code coverage result.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/FunctionOperatorUtils.java (hunk 1 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/FunctionOperatorUtils.java:81]: I'll refactor this part. We do already have comparison operators in queries, it would be good to reuse those. Some refactoring might be needed to make them usable at indexing time.

> Is there not a way to know the type of a PropertyValue ?

There is.

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/KeysetPaginationTest.java (hunk 2 above)
Comment (thomasmueller) [reply oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/KeysetPaginationTest.java:1]: These functions were added to better support keyset pagination, so it does make sense to add tests for that. But the current test doesn't use the functions, right. I'll change that.

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/FulltextRegexQueryTest.java (hunk 6 above)
Comment (thomasmueller) [reply oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/FulltextRegexQueryTest.java]: No, I committed this by mistake! I'll remove it.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/ExistsImpl.java (hunk 3 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/ExistsImpl.java:37]: I'm adding more tests in FilterTest.functionBasedIndex

File: oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/util/FunctionIndexProcessor.java
Comment (thomasmueller) [reply oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/util/FunctionIndexProcessor.java]: Right. I'll ensure we have very good test coverage here.

File: oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/util/FunctionIndexProcessor.java (hunk 7 above)
Comment (thomasmueller) [reply oak-search/src/main/java/org/apache/jackrabbit/oak/plugins/index/search/util/FunctionIndexProcessor.java]: I changed the code such that all new code is at the top, the rest is unchanged.

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/SQL2Parser.java [hunk 9]
```diff
@@ -740,8 +749,27 @@ private DynamicOperandImpl parseExpressionFunction(String functionName) throws P
...
+        } else if ("IF".equalsIgnoreCase(functionName)) {
+            DynamicOperandImpl condition = parseDynamicOperand();
+            read(",");
+            DynamicOperandImpl trueValue = parseDynamicOperand();
+            read(",");
+            DynamicOperandImpl falseValue = parseDynamicOperand();
+            op = factory.ifOperand(condition, trueValue, falseValue);
+        } else if ("EXISTS".equalsIgnoreCase(functionName)) {
+            op = factory.existsOperand(parseDynamicOperand());
+        } else if ("OP".equalsIgnoreCase(functionName)) {
```
Comment (bhabegger) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/query/SQL2Parser.java]: Do we have a test for this OP function ?

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/IfImpl.java (hunk 4 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/ast/IfImpl.java]: FilterTest.functionBasedIndex

File: oak-core/src/main/java/org/apache/jackrabbit/oak/query/SQL2Parser.java (hunk 9 above)
Comment (thomasmueller) [reply oak-core/src/main/java/org/apache/jackrabbit/oak/query/SQL2Parser.java]: I'll think yes, but I'll check

File: oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/KeysetPaginationTest.java (hunk 2 above)
Comment (thomasmueller) [reply oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/KeysetPaginationTest.java]: I have now updated the keyset pagination docs and test (WIP)
